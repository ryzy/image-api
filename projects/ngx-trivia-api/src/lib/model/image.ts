export interface ImageLinks {
  /**
   * HTML link to the image (browse'able from browser)
   */
  html: string;

  large: string;
  regular: string;
  small: string;
  thumb: string;
}

/**
 * Object present when license is different than PublicDomain
 */
export interface ImageAuthor {
  name?: string;
  url?: string;
  avatar?: string;
  twitter?: string;
}

export interface Image {
  id: string;
  caption: string;
  source: 'Google' | 'Unsplash';
  license: 'cc_publicdomain' | 'cc_attribute';
  location?: string;
  links: ImageLinks;
  author?: ImageAuthor;
}
