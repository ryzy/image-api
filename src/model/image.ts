export interface ImageLinks {
  html: string;
  full: string;
  raw: string;
  regular: string;
  small: string;
  thumb: string;
}
export interface ImageAuthor {
  source: 'Unsplash';
  name: string;
  url: string;
  avatar: string;
  twitter: string;
}

export interface Image {
  id: string;
  caption: string;
  location: string;
  links: ImageLinks;
  author: ImageAuthor;
}

export function unsplashToImage(data: any = {}): Image {
  return {
    id: data.id,
    caption: data.description,
    location: data.location && data.location.title,
    links: {
      html: data.links && data.links.html,
      ...data.urls,
    },
    author: {
      source: 'Unsplash',
      name: data.user && data.user.name,
      url: data.user && data.user.links && data.user.links.html,
      avatar:
        data.user && data.user.profile_image && data.user.profile_image.large,
      twitter: data.user && data.user.twitter_username,
    },
  };
}
