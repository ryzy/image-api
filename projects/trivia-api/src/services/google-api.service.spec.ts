import { Test, TestingModule } from '@nestjs/testing';
import { HttpModule, HttpService } from '@nestjs/common';
import { marbles } from 'rxjs-marbles';
import { of, throwError } from 'rxjs';

import { GoogleApiService, googleResultToImage } from './google-api.service';
import { mockGoogleImageRes } from '../../../../test/fixtures/google-image';
import { mockImage3 } from '../../../../test/fixtures/image';

describe('GoogleApiService', () => {
  let app: TestingModule;
  let http: HttpService;
  let service: GoogleApiService;

  beforeAll(async () => {
    app = await Test.createTestingModule({
      imports: [HttpModule],
      providers: [GoogleApiService],
    }).compile();

    http = app.get<HttpService>(HttpService);
    service = app.get<GoogleApiService>(GoogleApiService);
  });

  test('#googleResultToImage', () => {
    expect(googleResultToImage(undefined as any, 1)).toMatchSnapshot();
    expect(googleResultToImage(mockGoogleImageRes, 0)).toEqual(mockImage3);
  });

  describe('getImages', () => {
    it(
      'should get images',
      marbles((m) => {
        jest.spyOn(http, 'get').mockReturnValue(of({ data: { items: [mockGoogleImageRes] } }));
        m.expect(service.getImages({ q: 'some query' })).toBeObservable('(v|)', {
          v: [mockImage3],
        });
      }),
    );

    it(
      'should should throw error when no query',
      marbles((m) => {
        const err = new Error('missing query');
        jest.spyOn(http, 'get').mockReturnValue(throwError(err));
        m.expect(service.getImages({})).toBeObservable('#', undefined, {
          name: 'Error',
          message: {
            errors: ['Internal Server Error'],
            status: 500,
          },
        });
      }),
    );
  });
});
