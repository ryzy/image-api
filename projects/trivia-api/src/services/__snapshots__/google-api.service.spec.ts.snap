// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleApiService #googleResultToImage 1`] = `
Object {
  "caption": undefined,
  "id": "google-image-1",
  "license": "cc_publicdomain",
  "links": Object {
    "html": undefined,
    "large": undefined,
    "regular": undefined,
    "small": undefined,
    "thumb": undefined,
  },
  "source": "Google",
}
`;
